### PR TITLE
Redid handleMove and handleClick logic

### PR DESCRIPTION
**Pull request by gaganza**
Still need to do another pass to refactor logic into funcitons but this handle all cases

### CODE REVIEW

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 1]
```diff
@@ -120,6 +120,11 @@ export default class Game extends React.Component {
     return squares;
   }
 
+  // returns a boolean indication whether a move is valid
+  checkIfMoveValid(index) {
+    return this.state.squares[index].movable;
+  }
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: Lovely little function, super readable

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 2]
```diff
@@ -139,22 +144,30 @@ export default class Game extends React.Component {
...
-      error: ""
-    });
+
+    if (!this.checkIfMoveValid(index)) {
+      this.setState({
+        error: "Cannot move piece there"
+      });
+    } else {
+      pieceToMove.deselectPiece();
+      this.dehighlightMoves();
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: Nice, glad to see this was added

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 3]
```diff
@@ -139,22 +144,30 @@ export default class Game extends React.Component {
...
+      postMovePosition[this.state.selected] = null;
+      this.setState({
+        squares: postMovePosition,
+        selected: NOT_SELECTED,
+        capturedBlackPieces: capturedB,
+        capturedWhitePieces: capturedW,
+        turn: this.state.turn === WHITE ? BLACK : WHITE,
+        count: this.state.count + 1,
+        error: ""
+      });
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: this whole setState call could probably be it's own function

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 4]
```diff
@@ -163,55 +176,87 @@ export default class Game extends React.Component {
...
-      // Check whether or not that peice is already selected
-      if (this.state.squares[index].selected) {
-        newState[index].deselectPiece();
-        this.setState({
-          squares: newState,
-          selected: NOT_SELECTED,
-          error: ""
-        });
-      } else {
+    let newState = this.state.squares;
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: I feel like this is someone odd naming. `newState` is pointing to the a list of squares, right? Maybe there's a better name, like `newChessBoard` or something along that lines? At least to me, `state` seems to suggest a the current state of a single variable, and that isn't what `newState` seems to be in this case.

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 5]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+    // if the player currently has no piece selected
+    if (this.state.selected === NOT_SELECTED) {
+      // if the spot that was clicked was the current player's color
+      if (this.state.squares[index].player === this.state.turn) {
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: Maybe this check can be wrapped in a function, `checkIfPieceBelongsToCurrentPlayer()`? A mouthful, but gets the logic across more easily then then `this.state.squares[index].player === this.state.turn`

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 6]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+        this.checkPossibleMoves(index);
         newState[index].selectPiece();
         this.setState({
           squares: newState,
           selected: index,
           error: ""
         });
+      } else if (this.state.squares[index].player === this.opponentColor()) {
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: Why is there a function for `opponentColor()` but not `currentColor()`?

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 7]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+      }
+    } else {
+      // if the player currently has a piece selected
+      // check the spot that was clicked
+
+      // if the spot that was clicked was the current player's color
+      if (this.state.squares[index].player === this.state.turn) {
+        // if the spot selected is the current selected piece, then we just need to
+        // deselect the piece
+        if (this.state.selected === index) {
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]:
```suggestion
       if (selectedPieceIsCurrentSelectedPiece()) {
```
Thoughts?

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 8]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+          this.dehighlightMoves();
+          newState[index].deselectPiece();
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: These two lines seem to be logically related, and could probably be their own function, `deselectMovesFromCurrentAndNewBoard`

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 9]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+          this.setState({
+            squares: newState,
+            selected: NOT_SELECTED,
+            error: ""
+          });
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: This setState call could be it's own function, `deselectAllPiecesOnBoard(newState)`

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 10]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+        } else {
+          // if the spot selected is not the current selected piece throw an error,
+          // since you cannot capture your own piece
+          this.setState({
+            error: "Cannot cannot capture your own piece"
+          });
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: replace with `setStateError(""Cannot cannot capture your own piece")`?

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 11]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+        }
+      } else if (this.state.squares[index].player === this.opponentColor()) {
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: `isSelectPieceOpponents()` or some function like that?

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 12]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+        // if the spot that was clcked was the opposite player's color
+        this.handleMove(true, index);
+      } else if (this.state.squares[index].player === null) {
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: `isSelectedPieceEmptySpot()`

File: chess/frontend/gui/src/modules/Game/Game.jsx [hunk 13]
```diff
@@ -163,55 +176,88 @@ export default class Game extends React.Component {
...
+        // if the spot that was clicked was an empty spot
+        this.handleMove(false, index);
```
Comment (jacobchrismarsh) [inline chess/frontend/gui/src/modules/Game/Game.jsx]: What does 'true' and 'false' here represent? Maybe we can wrap that in a variable like `isSelectionValid = true` and then call `this.handleMove(isSelectionValid, index)`

File: chess/frontend/gui/src/modules/Game/Game.jsx (hunk 6 above)
Comment (gaganza) [reply chess/frontend/gui/src/modules/Game/Game.jsx]: Because currentColor is just this.state.player

Comment (jacobchrismarsh) [reply chess/frontend/gui/src/modules/Game/Game.jsx]: True, but what is more readable? If there is an `opponentColor()` function, one would assume there is a `currentColor()` or `currentPlayerColor()` function

**Pull request merged**